mesonbuild/modules/freedesktop.py

```diff
@@ -0,0 +1,206 @@
+# Copyright 2019 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for freedesktop related standards."""
+
+import configparser
+import os
+import typing as T
+import xml.etree.ElementTree as et
+
+from . import ExtensionModule, ModuleReturnValue
+from .. import mlog
+from ..build import Executable, Data
+from ..interpreterbase import permittedKwargs, Disabler
+from ..mesonlib import MesonException, File, FileMode, listify, unholder
+
+if T.TYPE_CHECKING:
+    from ..interpreter import ModuleState, ExecutableHolder
+
+    _T = T.TypeVar('_T')
+
+
+class FDOConfigParser(configparser.ConfigParser):
+
+    """Special case config parser that doesn't muck with capitaliation."""
+
+    def optionxform(self, value: '_T') -> '_T':
+        return value
+
+
+def optional_string_arg(kwargs: T.Dict[str, T.Any], kkey: str,
+                        values: T.Dict[str, str], vkey: str) -> None:
+        value = kwargs.get(kkey)
+        if value is not None:
+            if not isinstance(kkey, str):
+                raise MesonException('"{}" must be a string if defined'.format(kkey))
+            values[vkey] = value
+
+
+def optional_list_arg(kwargs: T.Dict[str, T.Any], kkey: str,
+                      values: T.Dict[str, str], vkey: str) -> None:
+        value = listify(kwargs.get(kkey, []))
+        if value:
+            if any(not isinstance(i, str) for i in value):
+                raise MesonException('"{}" must be a list of strings if defined'.format(kkey))
+            values[vkey] = ';'.join(value)
+
+
+def subelem_text(parent: et.Element, tag: str, text: str, **kwargs: T.Dict) -> et.Element:
+    e = et.SubElement(parent, tag, **kwargs)
+    e.text = text
+    return e
+
+
+class FreedesktopModule(ExtensionModule):
+
+    @permittedKwargs({'name', 'generic_name', 'icon', 'categories',
+                      'actions', 'install_name', 'mimetype', 'comment',
+                      'dbus'})
+    def desktop_file(self, state: 'ModuleState',
+                     args: T.Tuple[T.Union['ExecutableHolder', 'Disabler']],
+                     kwargs: T.Dict[str, T.Any]) -> ModuleReturnValue:
+        if len(args) != 1:
+            raise MesonException('Only one positinal argument allowed.')
+
+        if isinstance(args[0], Disabler):
+            return ModuleReturnValue(None, [])
```

Comment:
If an argument is a disabler the interpreter should not even call your method, are you sure this can happen?